packages/flutter/lib/src/material/tab_controller.dart

```diff
@@ -0,0 +1,201 @@
+          _indexIsChanging = false;
+          notifyListeners();
+        });
+    } else {
+      _indexIsChanging = true;
+      _animationController.value = _index.toDouble();
+      _indexIsChanging = false;
+      notifyListeners();
+    }
+  }
+
+  /// The index of the currently selected tab. Changing the index also updates
+  /// [previousIndex], sets the [animation]'s value, resets [indexIsChanging]
+  /// to false, and notifies listeners.
+  ///
+  /// To change the currently selected tab and play the [animation] use [animateTo].
+  int get index => _index;
+  int _index;
+  set index(int value) {
+    _changeIndex(value);
+  }
+
+  /// The index of the previously selected tab. Initially the same as [index].
+  int get previousIndex => _previousIndex;
+  int _previousIndex;
+
+  /// True while we're animating from [previousIndex] to [index].
+  bool get indexIsChanging => _indexIsChanging;
+  bool _indexIsChanging = false;
+
+  /// Immediately sets [index] and [previousIndex] and then plays the
+  /// [animation] from its current value to [index].
+  ///
+  /// While the animation is running [indexIsChanging] is true. When the
+  /// animation completes [offset] will be 0.0.
+  void animateTo(int value, { Duration duration: kTabScrollDuration, Curve curve: Curves.ease }) {
+    _changeIndex(value, duration: duration, curve: curve);
+  }
+
+  /// The difference between the [animation]'s value and [index]. The offset
+  /// value must be between -1.0 and 1.0.
+  ///
+  /// This property is typically set by the [TabBarView] when the user
+  /// drags left or right. A value between -1.0 and 0.0 implies that the
+  /// TabBarView has been dragged to the left. Similarly a value between
+  /// 0.0 and 1.0 implies that the TabBarView has been dragged to the right.
+  double get offset => _animationController.value - _index.toDouble();
+  set offset(double value) {
+    assert(value != null);
+    assert(value >= -1.0 && value <= 1.0);
+    assert(!indexIsChanging);
+    if (value == offset)
+      return;
+    _animationController.value = value + _index.toDouble();
+  }
+
+  @override
+  void dispose() {
+    _animationController.dispose();
+    super.dispose();
+  }
+}
+
+class _TabControllerScope extends InheritedWidget {
+  _TabControllerScope({
+    Key key,
+    this.controller,
+    this.enabled,
+    Widget child
+  }) : super(key: key, child: child);
+
+  final TabController controller;
+  final bool enabled;
+
+  @override
+  bool updateShouldNotify(_TabControllerScope old) {
+    return enabled != old.enabled || controller != old.controller;
+  }
+}
+
+/// The [TabController] for descendant widgets that don't specify one explicitly.
+class DefaultTabController extends StatefulWidget {
+  DefaultTabController({
+    Key key,
+    @required this.length,
+    this.initialIndex: 0,
+    this.child
+  }) : super(key: key);
+
+  /// The total number of tabs. Must be greater than one.
```

Comment:
Unless I'm misunderstanding something, the TabController constructor does assert about length:
```
class TabController extends ChangeNotifier {
  /// Creates an object that manages the state required by [TabBar] and a [TabBarView].
  ...
   ) {
    assert(length != null && length > 1);
    assert(initialIndex != null && initialIndex >= 0 && initialIndex < length);
  }
```